### PR TITLE
Preliminary work to wire to user secret facade to the secret service

### PR DESCRIPTION
**Pull request by wallyworld**
A new secret service is added with not yet implemented methods needed primarily to support the secrets client facade - this facade is used with the user secrets cli. The service methods are grouped and implemented in different files to avoid one big service.go file when done.

The secret service is a dummy implementation which panics when called. - the purpose here is to start moving the business logic from the facade layer to the service layer. The logic to save secret content to external backends is kept in the facade layer for now, until after the backend service work is landed. Logic to delete secret content can be moved to the service layer as we create the service with functions to allow admin access to the backend to do the deletes.

Some logic to handle auto prune is also moved from common., ready to be wired up later.
More logic is transplanted from the agent secret manager, also waiting to all be wired up.

Some existing calls from the facade layer to mongo state are removed since these will now be internal to the service. Test expectations are updated accordingly.

This work will need several PRs to replace the existing functionality - for now user secrets is broken until the service is implemented.

## QA steps

unit tests

## Links

**Jira card:** JUJU-5776

### CODE REVIEW

File: core/permission/access.go [hunk 1]
```diff
@@ -76,17 +76,19 @@ type ObjectType string
...
-	Cloud      ObjectType = "cloud"
-	Controller ObjectType = "controller"
-	Model      ObjectType = "model"
-	Offer      ObjectType = "offer"
+	Cloud       ObjectType = "cloud"
+	Controller  ObjectType = "controller"
+	Model       ObjectType = "model"
+	Offer       ObjectType = "offer"
+	Application ObjectType = "application"
+	Unit        ObjectType = "unit"
```
Comment (ycliuhw) [inline core/permission/access.go:84]: Is this change related to the secret?

File: apiserver/facades/client/secrets/secrets.go
Comment (ycliuhw) [inline apiserver/facades/client/secrets/secrets.go:1]: All the logic in this file will need to move to the service.

File: domain/secret/service/params.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	"github.com/juju/juju/core/permission"
+	"github.com/juju/juju/core/secrets"
+)
+
+// CreateSecretParams are used to create a secret.
+type CreateSecretParams struct {
+	UpdateSecretParams
+	Version int
+
+	// One owner must be non-nil.
```
Comment (ycliuhw) [inline domain/secret/service/params.go:19]: Probably add Validate() for this.

File: domain/secret/service/params.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+// UpdateSecretParams are used to update a secret.
+type UpdateSecretParams struct {
+	LeaderToken  leadership.Token
+	RotatePolicy *secrets.RotatePolicy
+	ExpireTime   *time.Time
+	Description  *string
+	Label        *string
+	Params       map[string]interface{}
+	Data         secrets.SecretData
+	ValueRef     *secrets.ValueRef
```
Comment (ycliuhw) [inline domain/secret/service/params.go:34]: And here cannot have both Data and ValueRef.

File: domain/secret/service/params.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	LeaderToken leadership.Token
+	Scope       permission.ID
+	Subject     permission.ID
+	Role        secrets.SecretRole
+}
+
+// ChangeSecretBackendParams are used to change the backend of a secret.
+type ChangeSecretBackendParams struct {
+	ValueRef *secrets.ValueRef
+	Data     secrets.SecretData
```
Comment (ycliuhw) [inline domain/secret/service/params.go:49]: Validate here.

File: domain/secret/service/params.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	UnitName        *string
+}
+
+// SecretAccessor represents an entity that can access a secret.
+// One of ApplicationName, UnitName, or ModelUUID must be non nil.
+type SecretAccessor struct {
+	ApplicationName *string
+	UnitName        *string
+	ModelUUID       *string
+}
```
Comment (ycliuhw) [inline domain/secret/service/params.go:65]: Validate here.

File: domain/secret/service/params.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	ValueRef *secrets.ValueRef
+	Data     secrets.SecretData
+}
+
+// SecretConsumer represents the consumer of a secret.
+// One of ApplicationName or UnitName must be non nil.
+type SecretConsumer struct {
+	ApplicationName *string
+	UnitName        *string
+}
```
Comment (ycliuhw) [inline domain/secret/service/params.go:57]: Validate here.

File: apiserver/facades/client/secrets/secrets_test.go
Comment (ycliuhw) [inline apiserver/facades/client/secrets/secrets_test.go:1]: We need to copy tests here to service later.

File: core/permission/access.go (hunk 1 above)
Comment (wallyworld) [reply core/permission/access.go:84]: Yes, we need to record permissions for secret grants

File: domain/secret/service/params.go (hunk 2 above)
Comment (wallyworld) [reply domain/secret/service/params.go:19]: That's all todo - this is just the skeleton for now. Only the minimal structs and not implemented service methods are added here.  The purpose of this PR is to start the facade rework. The service implementation comes next and is not done here.

File: domain/secret/service/params.go (hunk 3 above)
Comment (wallyworld) [reply domain/secret/service/params.go:34]: Right. Validation and business logic not added in this PR though.

**Review by ycliuhw — APPROVED** [review]
Thank you for this!

File: core/permission/access.go [hunk 7]
```diff
@@ -76,17 +76,19 @@ type ObjectType string
 
 // These values must match the values in the permission_object_type table.
```
Comment (manadart) [inline core/permission/access.go:77]: You have violated this comment.
